### DISCUSSION TITLE
Add support for package export for a plugin

### DISCUSSION THREAD

**Discussion by decebals** [pf4j/pf4j#627] [Ideas]
The idea is to add a new attribute in `PluginDescriptor`, with the name `Plugin-Export` or other good name, and to use that attribute in `PluginClassLoader`. The plugin class loader will not resolve the classes that are not in the list of packages exported.
Other details:
- this attribute is optional (all the classes are resolved/loaded by the plugin class loader if the attribute is not specified)
- the `*` character is allow (for example `mycompany.api.*, mycompany.utils`)

Practically this feature prevent the use of internal packages/classes in plugins (the implementation is hidden but of course only at RUNTIME)

**Comment by decebals** [comment]
A more easy and straightforward implementation is:
1. create a new `Export` annotation that is a package level annotation
2. put `@Export` on any package (via `package-info.java`) that you want to export
3. PF4J will use an annotation processor to create a "database" with all the packages exported by a plugin (the storage is `packages.ids` - similar with `extensions.idx`)

So, the single responsibility of plugin's developer is to put an `@Export` annotation on what package(s) he wants to export, and PF4J makes the rest of the job.

The functionality is similar with `Export-Package` from OSGI but it is much easier to use.

What do you think?

**Comment by wolframhaussig** [comment]
The more obvious way for a java developer would be the export annotation in the package-info.

What would be the expected outcame if I have classes `com.github.test.A` and `com.github.test.internals.B` when annotating package `com.github.test` with the Export annotation? Would the subpackages be exported too?

Maybe we could allow the Export annotation to be on packages and classes? This way, we could export whole packages and also single classes...
